### ISSUE TITLE
Proposal: Data classes

### ISSUE THREAD

**Issue by MichaelRFairhurst** [MichaelRFairhurst/wake-compiler#161]
So neither PHP nor java support properties on an interface.

This means I will have to generate getters/setters in java (hurting performance) and _get/_set methods in PHP (hurting performance more) and maybe most importantly just making interop more compilcated.

I realized that I will have the same problem in C++, as interfaces will mean multiple inheritance, and  multiple inheritance + members = diamond problem.

So I got to thinking, I don't like public members anyway, but I want to have them for their performance. But what if I made data classes?

Data classes would
- not be able to have methods 
- not be able to be implemented by other classes, only extended, and only by other data classes
- be able to be instantiated without a provider (!)
- be able to have _very testable_ methods via service provided methods (!)
- have no private members

Other classes would
- have no public members

Basically

    every Point:
        needs public Num x, public Num y;
        with scale = 1;

would be

    data Point:
        needs Num x, Num y;
        with scale = 1;

And you could do make them any time

    var Point(1, 2);

instead of

    var Point(1,2) from PointProvider

and you'd be able to still call methods on them, as long as they are on a service provider...

    every PointService:
        on Point
        allows getDistance() {
            return Point.x * Point.y * Point.scale;
        }

    every MyClass:
        needs PointService then {
            new Point(1, 2).getDistance();

Its an intriguing pairing of ideas. I don't like that its "more." But otherwise I like it a decent amount.

**Comment by MichaelRFairhurst** [comment]
Note that I was able to confirm that this can't really be done in java bytecode. The bytecode for accessing a field looks like so:

```
getfield "wake/your/module/Point/x" "I"
```

I for integer.

This works fine when we are dealing with our original "Point" class, but if I were to create a mock version, it _must_ also extend Point. If it does not, this field won't be found by the JVM.

Since means PointMock cannot both be a 'capable Point' and extend Mock (assuming Mock also has public fields), breaking the Wake spec.
